pkg/migration/migration.go

```diff
@@ -101,15 +113,195 @@ func (h *Handler) createStorageClusterIfAbsent(ds *appsv1.DaemonSet) (*corev1.St
+) error {
 	// TODO: Implement this
 	// 1. Backup existing specs
-	// 2. Migrate daemonset pods
-	// 3. Migrate components
-	// 4. Delete daemonset and components
+	// 2. Migrate components
+
+	nodeList := &v1.NodeList{}
+	if err := h.client.List(context.TODO(), nodeList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	nodes, err := h.markAllNodesAsPending(nodeList)
+	if err != nil {
+		return err
+	}
+
+	// Unblock operator reconcile loop to start managing the storagecluster
+	cluster.Status.Phase = constants.PhaseMigrationInProgress
+	if err := h.client.Status().Update(context.TODO(), cluster, &client.UpdateOptions{}); err != nil {
+		return err
+	}
+
+	if err := h.updateDaemonsetToRunOnPendingNodes(ds); err != nil {
+		return err
+	}
+
+	portworxNodes := sortedPortworxNodes(cluster, nodes)
+	for _, node := range portworxNodes {
+		nodeLog := logrus.WithField("node", node.Name)
+		nodeLog.Infof("Starting migration of portworx pod")
+
+		value := node.Labels[constants.LabelPortworxDaemonsetMigration]
+		if value == constants.LabelValueMigrationDone {
+			nodeLog.Infof("Portworx pod already migrated")
+			continue
+		} else if value == constants.LabelValueMigrationSkip {
+			nodeLog.Infof("Portworx pod migration skipped")
+			continue
+		}
+
+		if err := h.markMigrationAsStarting(node); err != nil {
+			return err
+		}
+
+		// Wait for daemonset pod to terminate, else it causes conflicts with
+		// the operator managed portworx pod
+		if err := h.waitForDaemonSetPodTermination(ds, node.Name, nodeLog); err != nil {
+			return err
+		}
+
+		if err := h.markMigrationAsInProgress(node); err != nil {
+			return err
+		}
+
+		// Wait until operator managed portworx pod is ready
+		if err := h.waitForPortworxPod(cluster, node.Name, nodeLog); err != nil {
+			return err
+		}
+
+		if err := h.markMigrationAsDone(node); err != nil {
+			return err
+		}
+
+		value = node.Labels[constants.LabelPortworxDaemonsetMigration]
+		if value == constants.LabelValueMigrationDone {
+			nodeLog.Infof("Portworx pod successfully migrated")
+		} else if value == constants.LabelValueMigrationSkip {
+			nodeLog.Infof("Portworx pod migration skipped")
+		}
+	}
+
+	logrus.Infof("Deleting portworx DaemonSet")
+	if err := h.deletePortworxDaemonSet(ds); err != nil {
+		return err
+	}
+
+	logrus.Infof("Removing migration label from all nodes")
+	if err := h.unmarkAllDoneNodes(); err != nil {
+		return err
+	}
+
 	return nil
 }
 
+func (h *Handler) waitForDaemonSetPodTermination(
+	ds *appsv1.DaemonSet,
+	nodeName string,
+	nodeLog *logrus.Entry,
+) error {
+	return wait.PollImmediate(podWaitInterval, daemonSetPodTerminationTimeout, func() (bool, error) {
+		node := &v1.Node{}
+		if err := h.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node); err != nil {
+			nodeLog.Errorf("Failed to get node. %v", err)
+			return false, nil
+		}
+		value := node.Labels[constants.LabelPortworxDaemonsetMigration]
+		if value == constants.LabelValueMigrationSkip {
```

Comment:
when we get here, it is probably too late to skip as it would be scaled down anyway. we may just check skip in the entry and not check here, and if it happens it would migrate anyway, what do you think?

Reply:
Actually skip does not skip the migration on that node. The daemonset pod is still deleted and the operator pod created. I don't think we need to support a case where user wants to keep some daemonset pods and some operator pods.
I added the skip to allow users to skip a node if it is not coming up for some known issue or may be an existing issue with their cluster. We don't want to get stuck at that node, nor do we want to proceed without user acknowledging the problem is known and we can skip the node. 
Basically the skip will allow the migration to proceed to a new node without worrying about whether pod comes up healthy or not.

Now the reason we want to skip from here is that if the user adds the skip flag later, we end up waiting for the timeout and then on the next restart we honor the skip. To make it react slightly faster, I added the skip check here.

Reply:
i was thinking about a scenario that when it is being migrated and user later added the skip annotation, we now have two pods as we dont wait here.

Reply:
As I said before, the skip does not stop the daemonset from deleting and operator from creating the pod. That still happens, it just skips the waiting on that node and proceeds with the migration.